packages/lib/AsyncActionQueue.ts

```diff
@@ -1,40 +1,49 @@
+import Logger from '@joplin/utils/Logger';
 import shim from './shim';
 
-export interface QueueItemAction {
-	(): void;
+export interface QueueItemAction<Context> {
+	(context: Context): void|Promise<void>;
 }
 
-export interface QueueItem {
-	action: QueueItemAction;
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Old code before rule was applied
-	context: any;
+export interface QueueItem<Context> {
+	action: QueueItemAction<Context>;
+	context: Context;
 }
 
 export enum IntervalType {
 	Debounce = 1,
 	Fixed = 2,
 }
 
+type CanSkipTaskHandler<Context> = (current: QueueItem<Context>, next: QueueItem<Context>)=> boolean;
+
+const logger = Logger.create('AsyncActionQueue');
+
 // The AsyncActionQueue can be used to debounce asynchronous actions, to make sure
 // they run in the right order, and also to ensure that if multiple actions are emitted
 // only the last one is executed. This is particularly useful to save data in the background.
 // Each queue should be associated with a specific entity (a note, resource, etc.)
-export default class AsyncActionQueue {
+export default class AsyncActionQueue<Context = void> {
 
-	private queue_: QueueItem[] = [];
+	private queue_: QueueItem<Context>[] = [];
 	private interval_: number;
 	private intervalType_: number;
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Old code before rule was applied
-	private scheduleProcessingIID_: any = null;
+	private scheduleProcessingIID_: ReturnType<typeof shim.setInterval>|null = null;
 	private processing_ = false;
 
 	public constructor(interval = 100, intervalType: IntervalType = IntervalType.Debounce) {
 		this.interval_ = interval;
 		this.intervalType_ = intervalType;
 	}
 
-	// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Old code before rule was applied
-	public push(action: QueueItemAction, context: any = null) {
+	// Determines whether an item can be skipped in the queue. Prevents data loss in the case that
+	// tasks that do different things are added to the queue.
+	private areItemsIdentical_: CanSkipTaskHandler<Context> = (_a, _b) => true;
+	public setCanSkipTaskHandler(callback: CanSkipTaskHandler<Context>) {
+		this.areItemsIdentical_ = callback;
+	}
```

Comment:
I don't know if tasks being identical should be the reason they can be skipped or not. If they are idempotent yes, but we can't know that unless the caller defines it.

Is it possible to add some "mode" parameter to the queue object that tells whether tasks can be skipped or not? In which case you setup the queue with one more or another when you create it. Or in your use case is there a mix of tasks that can and cannot be skipped in the same queue?

Reply:
> Or in your use case is there a mix of tasks that can and cannot be skipped in the same queue?

In file mirroring, file watcher, database, and calls to `fullSync` all use the same queue. The goals of using an `AsyncActionQueue` are:
1. Prevents watcher and database update events from being handled during a `fullSync`.
2. Prevents multiple full syncs from happening at the same time.
3. Skips repeated "updated" tasks for the same note.
   - A user changing the title of a note might produce several repeated updated events.

> Is it possible to add some "mode" parameter to the queue object that tells whether tasks can be skipped or not?

Currently the custom `setCanSkipTaskHandler` is used to:
- Skip idempotent database update listeners and fullSync tasks.
- Not skip repeated file system watcher events.

On reflection, this should also be possible with a deep (or even shallow?) equality check of neighboring tasks' `context` and callback function.

In this case, a `mode` parameter similar to the following would work:
```ts
enum MergeTaskMode {
    AlwaysSkip,
    SkipSameCallbackAndContext,
}
```

> If they are idempotent yes, but we can't know that unless the caller defines it.

Currently, `areItemsIdentical_` isn't well-named. Based on how it's used, `areNeighboringTasksIdempotent_` could be a better name.

Reply:
Ok thanks for clarifying. I think it indeed makes sense in this case to give more control over what can be skipped or not. You can probably leave the code as it is now, but just rename `areItemsIdentical_` to `canSkipTaskHandler_` to show that it's up to the caller to define what can be skipped or not, and then we don't need the complexity of knowing here if something is idempotent or not.